File: source_repo/src/parser/links.js
'use strict';

/**
 * Extract all links from Markdown body content.
 *
 * Categorizes links as:
 *  - "internal": relative paths (e.g., ./other.md, ../docs/file.md, file.md#anchor)
 *  - "external": absolute URLs (http://, https://)
 *  - "anchor": same-page anchors (#section)
 *  - "other": mailto:, ftp://, etc.
 *
 * Each link includes:
 *  - text: the display text
 *  - url: the raw URL
 *  - type: one of the categories above
 *  - line: 1-based line number
 *
 * @param {string} body
 * @returns {Array<{text: string, url: string, type: string, line: number}>}
 */
function extractLinks(body) {
  const links = [];
  const lines = body.split('\n');
  let inCodeBlock = false;

  // Matches [text](url) — standard inline Markdown links
  const linkPattern = /$$([^$$]*)\]\(([^)]+)\)/g;

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];

    if (line.trimStart().startsWith('```') || line.trimStart().startsWith('~~~')) {
      inCodeBlock = !inCodeBlock;
      continue;
    }

    if (inCodeBlock) {
      continue;
    }

    let match;
    linkPattern.lastIndex = 0;

    while ((match = linkPattern.exec(line)) !== null) {
      const text = match[1];
      const url = match[2].trim();
      const type = classifyLink(url);

      links.push({
        text,
        url,
        type,
        line: i + 1,
      });
    }
  }

  return links;
}

/**
 * Classify a URL into a link type.
 * @param {string} url
 * @returns {string}
 */
function classifyLink(url) {
  if (url.startsWith('#')) {
    return 'anchor';
  }
  if (url.startsWith('http://') || url.startsWith('https://')) {
    return 'external';
  }
  if (url.startsWith('mailto:') || url.startsWith('ftp://')) {
    return 'other';
  }
  return 'internal';
}

module.exports = { extractLinks, classifyLink };

File: golden_repo/mdforge/parser/links.py
"""Link extraction and classification."""

import re

LINK_PATTERN = re.compile('\\[([^\\]]*)\\]\\(([^)]+)\\)')


def extract_links(body: str) -> list[dict]:
    """Extract all links from Markdown body content."""
    links = []
    lines = body.split("\n")
    in_code_block = False

    for i, line in enumerate(lines):
        trimmed = line.lstrip()

        if trimmed.startswith("```") or trimmed.startswith("~~~"):
            in_code_block = not in_code_block
            continue

        if in_code_block:
            continue

        for match in LINK_PATTERN.finditer(line):
            text = match.group(1)
            url = match.group(2).strip()
            link_type = classify_link(url)

            links.append({
                "text": text,
                "url": url,
                "type": link_type,
                "line": i + 1,
            })

    return links


def classify_link(url: str) -> str:
    """Classify a URL into a link type."""
    if url.startswith("#"):
        return "anchor"
    if url.startswith("http://") or url.startswith("https://"):
        return "external"
    if url.startswith("mailto:") or url.startswith("ftp://"):
        return "other"
    return "internal"
